File: magpie/src/rotate_2d.f90

subroutine rotate_2d_scalar(x, y, dphi, xrot, yrot)

  ! Rotates a single point in 2D cartesian coordinates counter clockwise by
  ! dphi in radians.
  !
  ! Parameters
  ! ----------
  ! x, y : float
  !   Cartesian coordinates.
  ! dphi : float
  !   Counter-clockwise rotation.
  !
  ! Returns
  ! -------
  ! xrot, yrot : float
  !   Rotated x and y coordinates.

  implicit none

  ! Parameter declarations

  integer, parameter :: dp = kind(1.d0)
  real(kind=dp), intent(in) :: x, y, dphi
  real(kind=dp), intent(out) :: xrot, yrot

  ! Main

  xrot = cos(dphi)*x - sin(dphi)*y
  yrot = sin(dphi)*x + cos(dphi)*y

end subroutine rotate_2d_scalar


subroutine rotate_2d_array(x, y, dphi, xlen, xrot, yrot)

  ! Rotates an array of points in 2D cartesian coordinates counter clockwise by
  ! dphi in radians.
  !
  ! Parameters
  ! ----------
  ! x, y : array
  !   Cartesian coordinates.
  ! dphi : float
  !   Counter-clockwise rotation.
  !
  ! Returns
  ! -------
  ! xrot, yrot : array
  !   Rotated x and y coordinates.

  implicit none

  ! Parameter declarations

  integer, parameter :: dp = kind(1.d0)
  integer, intent(in) :: xlen
  real(kind=dp), intent(in) :: x(xlen), y(xlen), dphi
  real(kind=dp), intent(out) :: xrot(xlen), yrot(xlen)

  integer :: i

  ! Main

  do i=1, xlen

    call rotate_2d_scalar(x(i), y(i), dphi, xrot(i), yrot(i))

  end do

end subroutine rotate_2d_array

File: magpie/src/rotate_2d.py
import numpy as np
from numba import njit


@njit
def rotate_2d_scalar(x, y, dphi):
    """
    Rotates a single point (x, y) counter-clockwise by dphi radians.

    Parameters
    ----------
    x, y : float
        Cartesian coordinates.
    dphi : float
        Counter-clockwise rotation in radians.

    Returns
    -------
    xrot, yrot : float
        Rotated x and y coordinates.
    """
    xrot = np.cos(dphi) * x - np.sin(dphi) * y
    yrot = np.sin(dphi) * x + np.cos(dphi) * y
    return xrot, yrot


@njit
def rotate_2d_array(x, y, dphi):
    """
    Rotates an array of points in 2D counter-clockwise by dphi radians.

    Parameters
    ----------
    x, y : array
        Cartesian coordinates of the points.
    dphi : float
        Counter-clockwise rotation in radians.

    Returns
    -------
    xrot, yrot : array
        Rotated x and y coordinates of the points.
    """
    xrot = np.zeros_like(x)
    yrot = np.zeros_like(y)

    for i in range(len(x)):
        xrot[i], yrot[i] = rotate_2d_scalar(x[i], y[i], dphi)

    return xrot, yrot
